File: n-queen-problem/c++/4-queen-problem.cpp
#include <iostream>
using namespace std;
class position {
public:
    int row;
    int col;
};

position queenPosition[4];

int rowFirstQueen = 0;
int colFirstQueen = 0;
bool success = false;

int board[4][4] = {
    { 0, 0, 0, 0 },
    { 0, 0, 0, 0 },
    { 0, 0, 0, 0 }
};
bool diagonal(position queen, int row, int col)
{
    int count = 1;
    //downward right
    while (count < 4) {
        for (int i = queen.row + count; i < 4; i++) {
            for (int j = queen.col + count; j < 4; j++) {
                if (i == row || j == col) {
                    return false;
                }
            }
        }
        count++;
    }

    count = 1;
    //downward left
    while (count < 4) {
        for (int i = queen.row + count; i < 4; i++) {
            for (int j = queen.col - count; j >= 0; j--) {
                if (i == row || j == col) {
                    return false;
                }
            }
        }
        count++;
    }

    count = 1;
    //upward left
    while (count < 4) {
        for (int i = queen.row - count; i >= 4; i--) {
            for (int j = queen.col - count; j >= 0; j--) {
                if (i == row || j == col) {
                    return false;
                }
            }
        }
        count++;
    }

    count = 1;
    //upward right
    while (count < 4) {
        for (int i = queen.row - count; i >= 4; i--) {
            for (int j = queen.col + count; j < 4; j++) {
                if (i == row || j == col) {
                    return false;
                }
            }
        }
        count++;
    }

    return true;
}
bool checkAttack(int i, int j)
{
    for (int i = 0; i < 4; i++) {

        //same row
        if (queenPosition[i].row == i) {
            return false;
        }

        //same column
        if (queenPosition[i].col == j) {
            return false;
        }

        //same diagonal
        if (!diagonal(queenPosition[i], i, j)) {
            return false;
        }
    }
    return true;
}
void map()
{

    board[rowFirstQueen][colFirstQueen] = 1;
    queenPosition[0].row = rowFirstQueen;
    queenPosition[0].col = colFirstQueen;

    for (int queen = 1; queen < 4; queen++) {
        bool found = false;
        for (int i = rowFirstQueen; i < 4; i++) {
            for (int j = colFirstQueen; j < 4; j++) {
                if (checkAttack(i, j)) {
                    board[i][j] = 1;
                    queenPosition[queen].row = i;
                    queenPosition[queen].row = j;
                    found = true;
                }
            }
        }
        if (!found) {
            if (colFirstQueen < 4) {
                colFirstQueen++;
            }
            else {
                colFirstQueen = 0;
                rowFirstQueen++;
            }
            map();
        }
        else {
            for (int i = 0; i < 4; i++) {
                for (int j = 0; j < 4; j++) {
                    board[i][j];
                }
            }
        }
    }
}

int main()
{
    map();
    return 0;
}

File: n-queen-problem/javascript/4-queen-problem.js
console.log('started');

var queensPosition=[
    {
    row:'',
    col:''
},
{
    row:'',
    col:''
},
{
    row:'',
    col:''
},
{
    row:'',
    col:''
}];

var rowFirstQueen=0;
var colFirstQueen=0;
var globalcounter=0;

function map(){
    var board=[
        [0,0,0,0],
        [0,0,0,0],
        [0,0,0,0],
        [0,0,0,0]
        ];

    board[rowFirstQueen][colFirstQueen]=1;
    queensPosition[0].row=rowFirstQueen;
    queensPosition[0].col=colFirstQueen;
    var found=false;

    for( var queen=1;queen<4;queen++){
        found=false;
        queen:
        for(var i=rowFirstQueen;i<4;i++){
            for(var j=colFirstQueen+1;j<4;j++){
                if(checkAttack(i,j)){
                    console.log("placing",i+","+j);
                    board[i][j]=1;
                    queensPosition[queen].row=i;
                    queensPosition[queen].row=j;
                    found=true;
                    break queen;
                }
            }
        }
    }
        if(!found){
            console.log('not found');
            if(colFirstQueen<3){
                colFirstQueen++;
            }else{
                colFirstQueen=0;
                rowFirstQueen++;
            }
            if(rowFirstQueen<4&&colFirstQueen<4){
            map();
            }
        }else{
            console.log('solved',board);
        }
        console.log(board);
}

function checkAttack(i,j){
    var returnValue=true;
    queensPosition.forEach(function(queen){
        //same row
        if(queen.row){
        if(queen.row==i||queen.col==j){
            returnValue= false;
        }else{
            returnValue=diagonal(queen,i,j);
        }
    }
    });
    if(i===1&&j===3&&queensPosition[0].row===0&&queensPosition[0].col===0){
        console.log(JSON.stringify(queensPosition));
    }
    if(returnValue===true){
        
    }
    return returnValue;
}

function diagonal(queen,row,col){
    var count=1;
    var returnValue=true;
    
    //downward right
    while(count<4){
    for(var i=queen.row+count;i<4;i++){
        for(var j=queen.col+count;j<4;j++){
               if(i==row||j==col){
                returnValue= false;
            }
        }
    }
count++;
}

count=1;
//downward left
while(count<4){
for(var i=queen.row+count;i<4;i++){
    for(var j=queen.col-count;j>=0;j--){
        if(i==row||j==col){
            returnValue= false;
        }
    }
}
count++;
}

count=1;
//upward left
while(count<4){
for(var i=queen.row-count;i>=4;i--){
    for(var j=queen.col-count;j>=0;j--){
        if(i==row||j==col){
            returnValue= false;
        }
    }
}
count++;
}

count=1;
//upward right
while(count<4){
for(var i=queen.row-count;i>=4;i--){
    for(var j=queen.col+count;j<4;j++){
        if(i==row||j==col){
            returnValue= false;
        }
    }
}
count++;
}
if(row===1&&col===3&&queensPosition[0].row===0&&queensPosition[0].col===0){
    console.log(JSON.stringify(returnValue));
}
return returnValue;
}

map();
